lib/Sema/CSFix.cpp

```diff
@@ -759,6 +759,50 @@ IgnoreContextualType *IgnoreContextualType::create(ConstraintSystem &cs,
       IgnoreContextualType(cs, resultTy, specifiedTy, locator);
 }
 
+bool RemoveUnnecessaryCoercion::diagnose(Expr *root, bool asNote) const {
+  auto &cs = getConstraintSystem();
+  UnnecessaryCoercionFailure failure(root, cs, getFromType(), getToType(),
+                                     getLocator());
+  return failure.diagnoseAsError();
+}
+
+bool
+RemoveUnnecessaryCoercion::attempt(ConstraintSystem &cs,
+                                   Type fromType,
+                                   Type toType,
+                                   ConstraintLocatorBuilder locator) {
+  if (auto last = locator.last()) {
+    if (last->is<LocatorPathElt::ExplicitTypeCoercion>()) {
+      
+      auto expr = cast<CoerceExpr>(locator.getAnchor());
+      auto toTypeRepr = expr->getCastTypeLoc().getTypeRepr();
+      
+      // only diagnosing for left-side is a DeclRefExpr or a
+      // explicit/implicit coercion e.g. Double(1) as Double
+      if (!isa<ImplicitlyUnwrappedOptionalTypeRepr>(toTypeRepr) &&
+          (isa<DeclRefExpr>(expr->getSubExpr()) || isa<CoerceExpr>(expr->getSubExpr()))) {
+        
+        auto *locatorPtr = cs.getConstraintLocator(locator);
+        auto hasFix = llvm::any_of(cs.getFixes(),
+                                   [&locatorPtr](const ConstraintFix *fix) -> bool {
+          if (fix->getLocator() == locatorPtr)
+            return true;
+          // If the anchor expr already has a fix, don't add this.
+          return fix->getLocator()->getAnchor() == locatorPtr->getAnchor();
```

Comment:
Hmm I don't think we should need this, but if we're generating duplicate diagnostics then maybe something is going wrong elsewhere (...or is this expected?).

Reply:
Yeah, I'm looking at this now, you probably right. I'm suspecting I'm not getting the right locator in CSApply, so that's probably why is being called here with two different locator pointers for the same disjunction choice.